src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php

```diff
@@ -194,6 +202,90 @@ public function testVariantInheritance(): void
         static::assertEquals('parent/green', $urls[$ids->get('green')]);
     }
 
+    public function testNotBeingStateful(): void
+    {
+        $categoryIds = $this->getCategoryIds(2);
+
+        self::assertCount(2, $categoryIds, 'this is important for the test as you need more items to iterate for a context switch test');
+
+        /** @var \Generator $firstRun */
+        $firstRun = $this->seoUrlGenerator->generate(
+            $categoryIds,
+            'template first run',
+            $this->seoUrlRouteRegistry->findByRouteName(TestNavigationSeoUrlRoute::ROUTE_NAME),
+            $this->salesChannelContext->getContext(),
+            $this->salesChannelContext->getSalesChannel()
+        );
+        /** @var \Generator $secondRun */
+        $secondRun = $this->seoUrlGenerator->generate(
+            $categoryIds,
+            'template second run',
+            $this->seoUrlRouteRegistry->findByRouteName(TestNavigationSeoUrlRoute::ROUTE_NAME),
+            $this->salesChannelContext->getContext(),
+            $this->salesChannelContext->getSalesChannel()
+        );
+
+        /** @var SeoUrlEntity $url */
+        foreach ($firstRun as $url) {
+            static::assertSame('template first run', $url->getSeoPathInfo());
+
+            break;
+        }
+
+        // this changes the template of the twig state to second template
+        \next($secondRun);
+
+        /** @var SeoUrlEntity $url */
+        foreach ($firstRun as $url) {
+            static::assertSame('template first run', $url->getSeoPathInfo());
+        }
+    }
+
+    public function testErrorLogging()
+    {
+        $logger = new class () extends AbstractLogger
+        {
+            public $logs = [];
+
+            public function log($level, $message, array $context = array())
```

Comment:
can we typehint there maybe already for psr/log 3?

Reply:
and isn't there any already existing logger for this?

Reply:
Sure I can.

I don't know. I just need to capture some logs to compare them later. Tell me :D Do you have such test tooling?

Reply:
Wait, we can't go psr/log 3 because this is php 8.0 only. Or is shopware already 8.0 only?